source/extensions/common/tap/tap_matcher.h

```diff
@@ -298,6 +343,94 @@ class HttpResponseTrailersMatcher : public HttpHeaderMatcherBase {
   }
 };
 
+/**
+ * Base class for body matchers.
+ */
+class HttpBodyMatcherBase : public SimpleMatcher {
+public:
+  HttpBodyMatcherBase(const std::vector<MatcherPtr>& matchers) : SimpleMatcher(matchers) {}
+
+protected:
+  // Limit search to specified number of bytes.
+  // Value equal to zero means no limit.
+  uint32_t limit_{};
+};
+
+/**
+ * Context is used by HttpGenericBodyMatcher to:
+ * - track how many bytes has been processed
+ * - track patterns which have been found
+ * - store last several seen bytes of the HTTP body (when pattern starts at the end of previous body
+ *   chunk and continues at the beginning of the next body chunk)
+ */
+class HttpGenericBodyMatcherCtx : public MatcherCtx {
+public:
+  HttpGenericBodyMatcherCtx(std::list<std::string> patterns, size_t overlap_size)
+      : patterns_(patterns), overlap_size_(overlap_size) {
+    overlap_ = std::make_unique<char[]>(overlap_size_);
+  }
+  virtual ~HttpGenericBodyMatcherCtx() = default;
+
+  // List is initialized based on the Matcher's configuration.
+  // When data is passing through the matcher, found patterns are removed
+  // from the list. When all patterns have been found, the list is empty.
+  std::list<std::string> patterns_;
+  // Stores the length of the longest pattern.
+  size_t overlap_size_;
+  // Buffer to store the last bytes from previous body chunk(s).
+  // It will store only as many bytes as is the length of the longest
+  // pattern to be found minus 1.
+  // It is necessary to locate patterns which are spread across 2 or more
+  // body chunks.
+  std::unique_ptr<char[]> overlap_;
```

Comment:
My preference would be to replace all 3 members related to `overlap_` with `std::vector`. I think you can avoid all manual `mem*` operations. The code as it is correct however it is error prone and does not offer any way to check overflows/underflows the way std::vector could.

Reply:
That is great suggestion. I refactored the code and it made things simpler and cleaner. However, I cannot do all operations via vector's methods. When copying out of `buffer`, vector's internal data has to be accessed directly (vector::data method). The reason is that `buffer` is a wrapper around list of data slices which are not continuous. I could use `buffer::linearize` and use `vector::assign`, but linearize seems to be expensive.